New model upload format

mharvan opened the pull request:
Fixes #16 

Model format is compatible with
  https://docs.aws.amazon.com/deepracer/latest/developerguide/deepracer-troubleshooting-service-migration-errors.html

The current best way for local training seems to be:
1. Upload model to s3
2. Import model from s3 in AWS DeepRacer console.

mharvan committed:
New model upload format

Format is compatible with
  https://docs.aws.amazon.com/deepracer/latest/developerguide/deepracer-troubleshooting-service-migration-errors.html

The current best way for local training seems to be:
1. Upload model to s3
2. Import model from s3 in AWS DeepRacer console.

DarrenBro commented on upload-current.sh:
```diff
@@ -1,77 +1,89 @@
-
-  echo "latest checkpoint = "$CHECKPOINT
+  #echo "Checkpoint not supplied, checking for latest checkpoint"
+  LAST_CHECKPOINT=`cat $CHECKPOINT_FILE |jq ".last_checkpoint.name" | sed s/\"//g`
```
You shouldn't need LAST_CHECKPOINT, BEST_CHECKPOINT or even just CHECKPOINT would be enough.

mharvan replied:
I always use the command line option -b, so end up using BEST_CHECKPOINT.
But the script in Matt's repo was defaulting to last checkpoint, so I guess that others prefer to upload last checkpoint rather than best checkpoint.

mharvan replied:
As discussed on slack, we'll default to last checkpoint. Best checkpoint can be used with the cmd line option -b.

DarrenBro commented on upload-current.sh:
```diff
@@ -1,77 +1,89 @@
-VAR2=`echo $VAR1 |jq ".last_checkpoint.name = $CHECKPOINT"`
-VAR3=`echo $VAR2 |jq ".best_checkpoint.name = $CHECKPOINT"`
-echo $VAR3 >checkpoint/deepracer_checkpoints.json
+echo $CHECKPOINT > checkpoint/model/.coach_checkpoint
```
I know it's only an echo, but good to share that you no longer need .coach_checkpoint in s3 for models with the new update.

adam-aph commented on upload-current.sh:
```diff
@@ -1,77 +1,89 @@
+cp -v $MODEL_FILE checkpoint/model/
+cp -v $METADATA_FILE checkpoint/model/
+
+CHECKPOINT_FILES=$MODEL_DIR/${CHECKPOINT}*
```
It seems tha according to the spec we need current checkpoint and previous one (so 2 checkpoint sets) - so the previous checkpoint should be copied or simply all checkpoints (there are two sets per last and best anyways), i.e. `CHECKPOINT_FILES=$MODEL_DIR/*.ckpt.*`

mharvan replied:
It works for race submission with a single checkpoint. Does something break without 2 checkpoints?

If we really need multiple checkpoints then we should upload all checkpoints.
Uploading all checkpoints takes longer, so I would only upload them all if this is really needed.

adam-aph commented on upload-current.sh:
```diff
@@ -1,77 +1,89 @@
-  aws s3 cp $filename s3://$S3_BUCKET/$S3_PREFIX/model/
-done
+# Cleanup upload destination
+aws s3 rm --recursive s3://$S3_BUCKET/$S3_PREFIX/
```
this one is removing all other folders which makes AWS unhappy, change it to: `aws s3 rm s3://$S3_BUCKET/$S3_PREFIX/model --recursive`

mharvan replied:
I forgot to include the directory upload-template. So some files were missing and that was causing issues.

I would always clean up the whole prefix to ensure that only newly uploaded files are present. Otherwise, unknown old files could be affecting the import and your model.

adam-aph commented on upload-current.sh:
```diff
@@ -1,77 +1,89 @@
 
-tar -czvf ${CHECKPOINT}-checkpoint.tar.gz checkpoint/*
+# Upload files to s3
+aws s3 sync checkpoint/ s3://$S3_BUCKET/$S3_PREFIX/
```
here it should be just: `aws s3 sync checkpoint/model s3://$S3_BUCKET/$S3_PREFIX/model`

mharvan replied:
The goal is to always upload a complete model with all required files, not just the model files. That includes also reward_function.py and ip/hyperparameters.json.

mharvan replied:
The tar is needed to keep an archive of what was uploaded. Once a new best model is found, sagemaker deletes the old best model so you would no longer have a copy.

mharvan committed:
Template for new model upload format

Add missing directory with upload template. The template includes a
generic reward function and generic hyperparameters.
